// _ccms-sdk@4.3.0@ccms-sdk/interceptors/metadata.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CREDENTIAL_KEY_MAPPER = exports.REQUEST_WHITE_LIST = exports.USER_SESSION_AVAILABLE_TIME = exports.REQUEST_TOKEN_VALUE = exports.REQUEST_TOKEN_HEADER = exports.noop = exports.Date = undefined;

var _credentials = require('../credentials');

var credential = (0, _credentials.getRequestCredential)(); /**
                                                            * @author qix
                                                            * @homepage https://github.com/qixman/
                                                            * @since 2016-10-21
                                                            */
var Date = exports.Date = window.Date;
var noop = exports.noop = function noop() {};
var REQUEST_TOKEN_HEADER = exports.REQUEST_TOKEN_HEADER = 'Authorization';
var REQUEST_TOKEN_VALUE = exports.REQUEST_TOKEN_VALUE = function REQUEST_TOKEN_VALUE(value) {
  return 'Bearer ' + value;
};
// export const USER_SESSION_AVAILABLE_TIME = 30 * 60 * 1000;
var USER_SESSION_AVAILABLE_TIME = exports.USER_SESSION_AVAILABLE_TIME = credential && credential['expires_in'] ? credential['expires_in'] * 1000 / 2 : 30 * 60 * 1000;
var REQUEST_WHITE_LIST = exports.REQUEST_WHITE_LIST = [];
var CREDENTIAL_KEY_MAPPER = exports.CREDENTIAL_KEY_MAPPER = {
  accessToken: 'access_token',
  refreshToken: 'refresh_token',
  expireTime: 'expire_time'
};
